File: balancedbintree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def helper(root):
            if root == None:
                return [0, True]
            case1 = helper(root.left)
            case2 = helper(root.right)
            h = 1 + max(case1[0],case2[0])
            if case1[1] == False or case2[1] == False:
                return [h, False]
            elif abs(case1[0] - case2[0]) > 1:
                return [h, False]
            return [h, True]

        if root == None:
            return True
        z = helper(root)
        return z[1]

File: BalancedBinTree.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    
    boolean isBalancedTree = true;
    
    public boolean isBalanced(TreeNode root) {
        // base check
        if (root == null) return true;
        
        helper(root);
        
        return isBalancedTree;
    }
    
    private int helper(TreeNode root) {
        // base case
        if (root == null) return 0;
        
        // Logic
        int left = helper(root.left);
        int right = helper(root.right);
        
        if (Math.abs(left - right) > 1)
            isBalancedTree = false;
        
        return Math.max(left, right) + 1;
    }
}